src/lib.rs

```diff
@@ -166,25 +177,95 @@ pub enum PixelLayout {
     Unknown,
 }
 
+#[derive(Copy, Clone, Debug)]
+pub enum PlanarImageComponent {
+    Y,
+    U,
+    V,
+}
+
+impl std::convert::From<usize> for PlanarImageComponent {
+    fn from(index: usize) -> Self {
+        match index {
+            0 => PlanarImageComponent::Y,
+            1 => PlanarImageComponent::U,
+            2 => PlanarImageComponent::V,
+            _ => panic!("Invalid YUV index: {}", index),
+        }
+    }
+}
+
+impl std::convert::From<PlanarImageComponent> for usize {
+    fn from(component: PlanarImageComponent) -> Self {
+        match component {
+            PlanarImageComponent::Y => 0,
+            PlanarImageComponent::U => 1,
+            PlanarImageComponent::V => 2,
+        }
+    }
+}
+
+#[derive(Clone, Debug)]
+pub struct Plane(Picture, PlanarImageComponent);
+
+impl AsRef<[u8]> for Plane {
+    fn as_ref(&self) -> &[u8] {
+        let (width, height) = self.0.plane_data_geometry(self.1);
+        unsafe {
+            std::slice::from_raw_parts(
+                self.0.plane_data_ptr(self.1) as *const u8,
+                (width * height) as usize,
+            )
+        }
+    }
+}
+unsafe impl Send for Plane {}
+unsafe impl Sync for Plane {}
+
 #[derive(Copy, Clone, Debug)]
 pub struct BitsPerComponent(pub usize);
 
 impl Picture {
-    pub fn stride(&self, component: usize) -> i32 {
-        (*self.pic).stride[component] as i32
+    pub fn stride(&self, component: PlanarImageComponent) -> i32 {
+        let s = match component {
+            PlanarImageComponent::Y => 0,
+            _ => 1,
+        };
+        (*self.inner).pic.stride[s] as i32
+    }
+
+    pub fn plane_data_ptr(&self, component: PlanarImageComponent) -> *mut c_void {
+        let index: usize = component.into();
+        (*self.inner).pic.data[index]
+    }
+
+    pub fn plane_data_geometry(&self, component: PlanarImageComponent) -> (u32, u32) {
```

Comment:
Is it guaranteed that the first component of the tuple here is always the same as the stride, i.e. `pic.stride[component]`?

Reply:
No, the stride is used now.